core/trino-main/src/main/java/io/trino/operator/HashGenerator.java

```diff
@@ -15,22 +15,15 @@
 
 import io.trino.spi.Page;
 
-import static com.google.common.base.Preconditions.checkState;
-
 public interface HashGenerator
 {
     long hashPosition(int position, Page page);
 
     default int getPartition(int partitionCount, int position, Page page)
     {
         long rawHash = hashPosition(position, page);
-
-        // clear the sign bit
-        rawHash &= 0x7fff_ffff_ffff_ffffL;
-
-        int partition = (int) (rawHash % partitionCount);
-
-        checkState(partition >= 0 && partition < partitionCount);
-        return partition;
+        // This function reduces the 64 bit rawHash to [0, partitionCount) uniformly. It first reduces the rawHash to 32 bit
+        // integer x then normalize it to x / 2^32 * partitionCount to reduce the range of x from [0, 2^32) to [0, partitionCount)
```

Comment:
I think it should be `x * partitionCounbt / 2^32`

Could you run a simple loop for `partitionCount=3` to see how uniform it really is?

Reply:
I tested partition counts 3, 5, 7, 11, 128, 256. 
Given enough positions (1M) the stdev of the distribution does not exceed 1%

Reply:
Here is the code I used (for 100K positions and 0.02)
```
@Test
    public void testGetPartitionDistribution()
    {
        Random random = new Random(365354);
        HashGenerator hashGenerator = (position, page) -> random.nextLong();
        IntStream.of(3, 5, 7, 11, 128, 256).forEach(partitionCount ->
        {
            System.out.printf("partition count %d\n", partitionCount);
            for (int l = 0; l < 100; l++) {
                int[] partitions = new int[partitionCount];
                int positionCount = 1_00_0000;
                for (int i = 0; i < positionCount; i++) {
                    partitions[hashGenerator.getPartition(partitionCount, 0, null)]++;
                }

                Stats stats = Stats.of(partitions);
                double stdevPercent = stats.populationStandardDeviation() / stats.mean();
                Assertions.assertThat(stdevPercent)
                        .describedAs(String.format("min %s, max %s, mean %s, stdev %s, stddev/mean %s\n%s",
                                stats.min(), stats.max(), stats.mean(), stats.populationStandardDeviation(), stdevPercent, Arrays.toString(partitions)))
                        .isLessThan(0.02);
            }
        });
    }
```